### PR TITLE
[release-2.2] cherry-pick: Fix cluster-api-provider-last-applied-tags-on-volumes to recalculate the value on each reconcile

### PR DESCRIPTION
**Pull request by Ankitasw**
**What type of PR is this?**
/kind bug

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
cherry-pick of #4531 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Fixed a bug with the `sigs.k8s.io/cluster-api-provider-aws-last-applied-tags` annotation that could cause excessive metadata values on AWSMachines. 
```

### CODE REVIEW

**Comment by Ankitasw** [comment]
cc @cnmcavoy

**Review by cnmcavoy — APPROVED** [review]

**Comment by Ankitasw** [comment]
cc @Skarlso @richardcase for LGTM

File: controllers/awsmachine_controller.go [hunk 1]
```diff
@@ -1111,14 +1112,15 @@ func (r *AWSMachineReconciler) indexAWSMachineByInstanceID(o client.Object) []st
...
+func (r *AWSMachineReconciler) ensureStorageTags(ec2svc services.EC2Interface, instance *infrav1.Instance, machine *infrav1.AWSMachine, additionalTags map[string]string) {
+	prevAnnotations, err := r.machineAnnotationJSON(machine, VolumeTagsLastAppliedAnnotation)
 	if err != nil {
 		r.Log.Error(err, "Failed to fetch the annotations for volume tags")
 	}
+	annotations := make(map[string]interface{}, len(instance.VolumeIDs))
 	for _, volumeID := range instance.VolumeIDs {
 		if subAnnotation, ok := annotations[volumeID].(map[string]interface{}); ok {
-			newAnnotation, err := r.ensureVolumeTags(ec2svc, aws.String(volumeID), subAnnotation, machine.Spec.AdditionalTags)
+			newAnnotation, err := r.ensureVolumeTags(ec2svc, aws.String(volumeID), subAnnotation, additionalTags)
```
Comment (philjb) [inline controllers/awsmachine_controller.go:1123]: the other side of this else was _also_ supposed to be changed if https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4512 was pulled in fully. This means release 2.2.4 contains some unapproved combination of code.

see https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4512/files#r1369368660

**Pull request merged**